File: src/main/java/org/teamvoided/transition/mappings/Mappings.java
package org.teamvoided.transition.mappings;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import org.teamvoided.transition.api.misc.MapCodecs;

import java.util.List;
import java.util.Map;

public record Mappings(Map<String, Map<String, String>> mappings) {

    public static final Codec<Mappings> CODEC = RecordCodecBuilder.create(instance -> instance
            .group(
                    Codec.unboundedMap(Codec.STRING, Codec.unboundedMap(Codec.STRING, Codec.STRING)).fieldOf("mappings").forGetter(Mappings::mappings)
            )
            .apply(instance, Mappings::new)
    );
}

File: src/main/kotlin/org/teamvoided/transition/mappings/Mappings.kt
package org.teamvoided.transition.mappings

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import org.teamvoided.transition.api.misc.MapCodecs

@JvmRecord
data class Mappings(
    @JvmField val oldNamespaces: MutableList<String>, @JvmField val oldToNewPaths: MutableMap<String, String>,
) {
    companion object {
        val CODEC: Codec<Mappings> =
            RecordCodecBuilder.create {
                it
                    .group(
                        Codec.STRING.listOf()
                            .optionalFieldOf("oldNamespaces", mutableListOf())
                            .forGetter(Mappings::oldNamespaces),
                        MapCodecs.codec(
                            Codec.pair(Codec.STRING.fieldOf("old").codec(), Codec.STRING.fieldOf("new").codec())
                        )
                            .optionalFieldOf("oldToNewPaths", mutableMapOf())
                            .forGetter(Mappings::oldToNewPaths)
                    )
                    .apply(it, ::Mappings)
            }
    }
}